lib/mllp/client.ex

```diff
@@ -0,0 +1,444 @@
+      {:ok, reply} ->
+        telemetry(:received, %{response: reply}, state)
+        maybe_verify_reply(state, message, reply, opts)
+
+      {:error, _} = err ->
+        err
+    end
+  end
+
+  defp maybe_receive_reply(_state, _msg, _opts) do
+    {:ok, :noop}
+  end
+
+  defp telemetry(_event_name, _measurements, %State{telemetry_module: nil} = _metadata) do
+    :ok
+  end
+
+  defp telemetry(event_name, measurements, %State{telemetry_module: telemetry_module} = metadata) do
+    telemetry_module.execute([:sender, event_name], add_timestamps(measurements), metadata)
+  end
+
+  defp add_timestamps(measurements) do
+    measurements
+    |> Map.put(:monotonic, :erlang.monotonic_time())
+    |> Map.put(:utc_datetime, DateTime.utc_now())
+  end
+
+  defp validate_options(opts) do
+    Map.get(opts, :address) || raise "No server address provided to connect to!"
+    Map.get(opts, :port) || raise "No server port provdided to connect to!"
+    opts
+  end
+
+  @default_opts %{
+    telemetry_module: MLLP.DefaultTelemetry,
+    tls_opts: [],
+    socket_opts: []
+  }
+
+  @default_send_opts %{
+    expect_reply: true,
+    timeout: 5000,
+    wrap_message: true,
+    reply_timeout: :infinity,
+    verify_reply: true
+  }
+
+  defp maybe_set_default_options(opts) do
+    socket_module = if opts.tls == [], do: TCP, else: TLS
+
+    send_opts =
+      opts
+      |> Map.take(Map.keys(@default_send_opts))
+
+    send_opts = Map.merge(@default_send_opts, send_opts)
+
+    opts
+    |> Map.merge(@default_opts)
+    |> Map.put_new(:tcp, socket_module)
+    |> Map.put(:pid, self())
+    |> Map.put(:tls_opts, opts.tls)
+    |> Map.put(:send_opts, send_opts)
+  end
+
+  def put_socket_address(%{address: address, port: port} = opts) do
+    Map.put(opts, :socket_address, "#{format_address(address)}:#{port}")
+  end
+
+  defp format_address(address) when is_list(address) or is_atom(address) or is_binary(address) do
+    to_string(address)
+  end
+
+  defp format_address(address), do: :inet.ntoa(address)
+
+  defp maybe_wrap_message(message, %{wrap_message: true}) do
+    MLLP.Envelope.wrap_message(message)
+  end
+
+  defp maybe_wrap_message(message, _), do: message
+
+  defp maybe_verify_reply(_, _, raw_ack, %{verify_reply: false}), do: {:ok, raw_ack}
+
+  defp maybe_verify_reply(_state, raw_message, raw_ack, %{verify_reply: true}) do
+    verify_ack(raw_ack, raw_message)
+  end
+
+  defp verify_ack(raw_ack, raw_message) do
+    ack = Envelope.unwrap_message(raw_ack)
+    unwrapped_message = Envelope.unwrap_message(raw_message)
+    Ack.verify_ack_against_message(unwrapped_message, ack)
+  end
+
+  defp new_error(context, error) do
+    %MLLP.Client.Error{
+      reason: error,
+      context: context,
+      message: format_error(error)
+    }
+  end
+
+  def format_error({:tls_alert, _} = err) do
+    to_string(:ssl.format_error({:error, err}))
+  end
+
+  def format_error(:closed), do: "connection closed"
```

Comment:
Are these formatting for logging purpose. Can we not simply do `inspect` on error?

Reply:
Yes,  for logging purposes.